### PR TITLE
Unable to publish/preview new assets when quick publish from tools config page editor

### PR DESCRIPTION
**Pull request by srikanthgurram-17**
This is to solve the issue https://github.com/wcm-io/io.wcm.caconfig.editor/issues/47.

- Right now the ConfigurationReferenceProvider is called when we try to publish the page using tools --> config page 
- The `ConfigurationReferenceProvider` class is responsible to identify all the ca config references in a page
- In future we need to support all the `assets type` referenced in ca config as well as `ca config type` reference we need to provide asset references which are outdated or newly added.
- I have updated this class so that it finds both types.

### CODE REVIEW

**Comment by srikanthgurram-17** [comment]
@stefanseifert  gentle reminder, As I am away after this Friday until next year to do any changes.

File: src/main/java/io/wcm/caconfig/extensions/references/impl/ConfigurationReferenceProvider.java [hunk 1]
```diff
@@ -200,4 +210,54 @@ private static String getType() {
...
+        Resource assetResource = resourceResolver.getResource((String) propertyValue);
+        if (isAssetReference(assetResource)) {
+          assetReferences.add(new com.day.cq.wcm.api.reference.Reference(DamConstants.ACTIVITY_TYPE_ASSET, assetResource.getName(), assetResource, getAssetLastModified(assetResource)));
+        }
+      }
+    }
+  }
+
+
+  private static long getAssetLastModified(Resource assetResource) {
```
Comment (stefanseifert) [inline src/main/java/io/wcm/caconfig/extensions/references/impl/ConfigurationReferenceProvider.java:241]: the `com.day.cq.dam.api.Asset` interface has a getLastModified() method, please use that.

File: src/main/java/io/wcm/caconfig/extensions/references/impl/ConfigurationReferenceProvider.java [hunk 2]
```diff
@@ -200,4 +210,54 @@ private static String getType() {
...
+   * Adds asset references from the given page to the provided list of asset references.
+   * This method checks the properties of the `jcr:content` node of the modified page for any
+   * references to assets in the `/content/dam/` path. If such references are found, they are
+   * added to the list of asset references.
+   *
+   * @param assetReferences The list to which asset references will be added.
+   * @param configPage The page whose properties are checked for asset references.
+   * @param resourceResolver The resource resolver used to resolve asset paths.
+   */
+  private void AddAssetReferencesOfaResource(List<com.day.cq.wcm.api.reference.Reference> assetReferences, Page configPage, ResourceResolver resourceResolver) {
```
Comment (stefanseifert) [inline src/main/java/io/wcm/caconfig/extensions/references/impl/ConfigurationReferenceProvider.java:223]: this method bypassed the context-aware configuration APIs and directly goes to the configuration page resource and reads all properties of the `jcr:content` node of the configuration page to check for any strings starting with `/content/dam`. this ignores a few important facts:
* the actual persistence of the configuration properties is up to the persistence strategy, no assumptions about that should be done here. so the configuration data should only be accessed via the caaconfig APIs. we have the "ConfigurationMetadata" in place for each configuration name checked in this class, so the actual defined property metadata is in place.
* esp. this method ignores the fact, that a flat list of properties is only the simplest use case. in case of nested configurations, configuration collections etc. the storage is different and more structures need to be checked - all using the caconfig APIs. inheritance and overrides might come into play as well.

to get this right is tricky, we need to cover all edge cases with nested, configuration lists etc. in the unit tests. also, this might come with an performance impact as we have to traverse all actual present and potentially present configuration data on all inheritance hierarchy levels.

i can help you at this point, but this may required a bit time.

**Review by stefanseifert — CHANGES_REQUESTED** [review]
thanks for the contribution!

there are some sonar findings (which unfortunately are not visible for external PRs directly in the PR comments), please have a look at https://sonarcloud.io/summary/new_code?id=wcm-io_io.wcm.caconfig.extensions&branch=feature%2Fassetreferences-publishing-support

overall this is the right approach, but there are some open issues.

File: src/main/java/io/wcm/caconfig/extensions/references/impl/ConfigurationReferenceProvider.java (hunk 2 above)
Comment (srikanthgurram-17) [reply src/main/java/io/wcm/caconfig/extensions/references/impl/ConfigurationReferenceProvider.java]: Hmm Thanks for explaining, 

- I see we have two different strategies OOTB from wcm.io, and we opted for page persistence strategy beginning of project start.  
- As per the use case I was aware of page level strategy and never get a chance to work with direct configuration api directly in the past (never setup by myself ). 
- Would be great if you can help here on top of my approach to cover in such a way that it will solve irrespective of persistence strategy. 

Best Regards
Srikanth Gurram

Comment (stefanseifert) [reply src/main/java/io/wcm/caconfig/extensions/references/impl/ConfigurationReferenceProvider.java]: will do - the AEM page persistence strategy is definitely the best/recommended one - but still users may pick different ones or event implement their own ones.

Comment (stefanseifert) [reply src/main/java/io/wcm/caconfig/extensions/references/impl/ConfigurationReferenceProvider.java]: after thinking more about it using the Configuration Management API for this is too complex, esp, as currently the reference provider implementation already bypasses it in a way, as it checks on resource level for any actual content pages in the configuration resource tree (which is fine for the goal). but we still need to cover the edge cases with nested configuration, configuration collections etc., will add that.

**Comment by stefanseifert** [comment]
i've refactored the code to cover more edge cases:
* detect asset references in configuration collections and nested configurations (stored as nested resources)
* detect asset references not only in string properties, but also in string array properties
* de-duplicate list of collected assets in case the same asset is references multiple times

i also moved the actual reference detection logic to a separate class to not clutter the main class too much, and make the edge cases testable independent of the other (complex) parts of the reference provider.

additionally, i added a new OSGi configuration option which allows to enable the detection of asset, but is disabled by default. so, we can test it a bit if it has performance impact, before enable it for everyone maybe in a later release.

**Review by stefanseifert — APPROVED** [review]
